cache_flat_mutation_reader.hh

```diff
@@ -642,10 +657,38 @@ void cache_flat_mutation_reader::copy_from_cache_to_buffer() {
     }, _read_context.is_reversed()) == stop_iteration::no) {
         return;
     }
-    // We add the row to the buffer even when it's full.
-    // This simplifies the code. For more info see #3139.
+
     if (_next_row_in_range) {
-        add_to_buffer(_next_row);
+        bool remove_row = false;
+
+        if (_state == state::reading_from_cache && !_next_row.dummy()) { // maybe we'd better use at_a_row() instead of dummy() here?
+            deletable_row& row = _next_row.latest_row();
+            tombstone base_tombstone = _snp->version()->partition().range_tombstone_for_row(*_schema, _next_row.key());
+
+            if (!row.is_live(*_schema, column_kind::regular_column, base_tombstone, _read_time) && // should we really call is_live here?
+                row.deleted_at().max_deletion_time() < _gc_before) {
+                remove_row = true;
```

Comment:
!row.is_live() is not the correct check. Row can be not live but still have not expired tombstones inside cells, which we must not drop.

The proper way to do it is to compact it using row.compact_and_expire() like mutation_compactor does, and then drop the row only if it is empty().

Reply:
you mean something like following?
```
bool expired(tombstone t) {
    return (t && t < _gc_before);
}

if (expired(range_tombstone) || expired(row_tombstone)) {
    if (!row.compact_and_expire(...)) {
        remove_row = true;
    }
}
```
anyway in some cases compact_and_expire (looks like expensive operation a little) will be called twice for each row: in `cache_flat_mutation_reader` and `mutation_compactor` and it makes me feel embarrassed

Reply:
Something like that, except that remove_row should not depend on the result of compact_and_expire(), but row.empty(). compact_and_expire() returns "is live". !is_live doesn't mean that we can remove the row, because it may still contain non-expired tombstones. Only when it contains no information after compaction, so empty(), can we remove it.

Reply:
```
    if (!row.compact_and_expire(...)) {
        remove_row = row.empty();
    }
```